### ISSUE TITLE
Support custom process group backends

### ISSUE THREAD

**Issue by tchaton** [PyTorchLightning/pytorch-lightning#11725]
### Motivation

Find the link there: https://github.com/facebookresearch/fairring

<img width="1069" alt="Screenshot 2022-02-03 at 19 21 26" src="https://user-images.githubusercontent.com/12861981/152413873-6bc56164-4c3d-4724-8d63-ec71613ae9cf.png">

<img width="678" alt="Screenshot 2022-02-03 at 19 20 26" src="https://user-images.githubusercontent.com/12861981/152413747-5495dda1-c38a-489e-9c29-381bf807cc85.png">

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @rohitgr7 @akihironitta

**Comment by ananthsub** [comment]
I've thought about this as well. IMO, the process group backend should be a constructor argument on the relevant strategies. 

The logic here is pretty hidden, especially with the environment variable. Making this a constructor argument would be much more explicit: https://github.com/PyTorchLightning/pytorch-lightning/blob/478c6085dfc2163f13dd36fd9fcdb8c1f12f5f54/pytorch_lightning/strategies/parallel.py#L101-L106

**Comment by daniellepintz** [comment]
I can work on this! (Making process group backend a constructor argument and then integrating with fairring)

**Comment by awaelchli** [comment]
The previous mechanism with the env var had the advantage of making the code hardware agnostic, i.e., you could set `PL_TORCH_DISTRIBUTED_BACKEND="gloo"` as fixed and freely choose the devices and hardware though e.g. CLI. Now, you would need to change the code and instantiate the strategy yourself, or register the custom strategy in the Registry.
I hope existing users will not be impacted by this.

**Comment by ananthsub** [comment]
@awaelchli - if there is pushback from users regarding the enviroment variable, we can always undeprecate it and continue supporting it. At the very least, this logic is now isolated to a single utility function under distributed.py
